packages/controller-utils/src/create-service-policy.test.ts

```diff
@@ -0,0 +1,2703 @@
+
+        await policy.execute(() => mockService());
+
+        expect(onDegraded).not.toHaveBeenCalled();
+      });
+
+      it('calls the onDegraded callback once if the service execution time beyond the threshold', async () => {
+        const degradedThreshold = 2000;
+        const delay = degradedThreshold + 1;
+        const mockService = jest.fn(() => {
+          return new Promise((resolve) => {
+            setTimeout(() => resolve({ some: 'data' }), delay);
+          });
+        });
+        const onDegraded = jest.fn();
+        const policy = createServicePolicy({ degradedThreshold, onDegraded });
+
+        const promise = policy.execute(() => mockService());
+        clock.tick(delay);
+        await promise;
+
+        expect(onDegraded).toHaveBeenCalledTimes(1);
+      });
+    });
+  });
+
+  describe('wrapping a service that always fails', () => {
+    describe(`using the default max retries (${DEFAULT_MAX_RETRIES})`, () => {
+      it(`calls the service a total of ${
+        1 + DEFAULT_MAX_RETRIES
+      } times, delaying each retry using a backoff formula`, async () => {
+        const error = new Error('failure');
+        const mockService = jest.fn(() => {
+          throw error;
+        });
+        const policy = createServicePolicy();
+        // Each retry delay is randomized using a decorrelated jitter formula,
+        // so we need to prevent that
+        jest.spyOn(Math, 'random').mockReturnValue(0);
+
+        const promise = policy.execute(() => mockService());
+        // These values were found by logging them
+        clock.tickAsync(0).catch(() => {
+          // ignore any errors - adding to the promise queue is enough
+        });
+        clock.tickAsync(176.27932892814937).catch(() => {
+          // ignore any errors - adding to the promise queue is enough
+        });
+        clock.tickAsync(186.8886145345685).catch(() => {
+          // ignore any errors - adding to the promise queue is enough
+        });
+        try {
+          await promise;
+        } catch {
+          // ignore the error
+        }
+
+        expect(mockService).toHaveBeenCalledTimes(1 + DEFAULT_MAX_RETRIES);
+      });
+
+      describe(`using the default max number of consecutive failures (${DEFAULT_MAX_CONSECUTIVE_FAILURES})`, () => {
+        it('throws what the service throws', async () => {
+          const error = new Error('failure');
+          const mockService = jest.fn(() => {
+            throw error;
+          });
+          const policy = createServicePolicy();
+
+          const promise = policy.execute(() => mockService());
+          clock.runAllAsync().catch(() => {
+            // ignore any errors - adding to the promise queue is enough
+          });
+
+          await expect(promise).rejects.toThrow(error);
+        });
+
+        it('does not call the onBreak callback, since the max number of consecutive failures is never reached', async () => {
```

Comment:
I am realizing that this is probably not what would happen in production code. If the policy is intended to be reused across multiple executions, then eventually `onBreak` _would_ get called once the _total_ number of attempts exceeds the max number of consecutive failures. The vast majority of these tests only use `policy` once, but I guess I could make an exception for this test? Curious to know your thoughts.